File: scraper-engine/pandascore_engine.py
import os
import requests
from dotenv import load_dotenv

# Load Env
load_dotenv('.env.local')
PANDASCORE_KEY = os.getenv('PANDASCORE_API_KEY', '').strip('"').strip("'")
SUPABASE_URL = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
SUPABASE_KEY = os.getenv('NEXT_PUBLIC_SUPABASE_ANON_KEY')

# Headers
ps_headers = {"Authorization": f"Bearer {PANDASCORE_KEY}", "Accept": "application/json"}
sb_headers = {"apikey": SUPABASE_KEY, "Authorization": f"Bearer {SUPABASE_KEY}", "Content-Type": "application/json", "Prefer": "return=representation"}

GAMES = ['valorant', 'csgo', 'dota-2']

def supabase_request(method, endpoint, payload=None):
    if not SUPABASE_URL or not SUPABASE_KEY:
        print("❌ Supabase credentials missing.")
        return None
        
    url = f"{SUPABASE_URL}/rest/v1/{endpoint}"
    try:
        if method == "GET":
            response = requests.get(url, headers=sb_headers)
        elif method == "POST":
            response = requests.post(url, headers=sb_headers, json=payload)
        
        if response.status_code >= 400:
            err_msg = response.text[:200].replace('\n', ' ')
            print(f"Supabase API Error ({response.status_code}): {err_msg}")
            return None
            
        return response.json() if response.text else []
    except Exception as e:
        print(f"Request Error: {e}")
        return None

def fetch_and_sync_matches():
    print("Starting PandaScore Engine...")
    
    # 1. Map our database game IDs
    print("Fetching Game IDs from Supabase...")
    games_data = supabase_request("GET", "games?select=id,slug")
    if games_data is None:
        print("❌ Could not fetch games from Supabase. Exiting.")
        return
        
    db_games = {g['slug'].replace('cs2', 'csgo'): g['id'] for g in games_data}
    print(db_games)

    for game in GAMES:
        if game not in db_games:
            continue
            
        game_id = db_games[game]
        print(f"\n--- Processing {game.upper()} ---")
        
        # Fetch Upcoming Matches
        url = f"https://api.pandascore.co/{game}/matches/upcoming?per_page=10"
        response = requests.get(url, headers=ps_headers)
        
        if response.status_code != 200:
            print(f"❌ Failed to fetch {game} matches: {response.status_code}")
            continue
            
        matches = response.json()
        print(f"Found {len(matches)} upcoming matches.")

        for match in matches:
            # 1. Ensure Tournament exists
            t = match['tournament']
            l = match.get('league') or {}
            
            tourney_data = {
                "name": f"{l.get('name', '')} {t.get('name', '')}".strip() or match.get('name', 'Unknown Tournament'),
                "slug": t.get('slug', str(t.get('id', match['id']))),
                "game_id": game_id,
                "tier": t.get('tier', 'c').capitalize(),
                "prize_pool": 0, # PandaScore free tier doesn't always have this parsed cleanly
                "start_date": t.get('begin_at'),
                "end_date": t.get('end_at'),
                "region": t.get('region', l.get('name', 'Global')),
                "status": "upcoming"
            }
            
            t_record = supabase_request("POST", "tournaments?on_conflict=slug", payload=tourney_data)
            if not t_record:
                t_record = supabase_request("GET", f"tournaments?slug=eq.{tourney_data['slug']}&select=id")
                
            if not t_record or (isinstance(t_record, list) and len(t_record) == 0):
                print(f"⚠️ Could not find or create tournament: {tourney_data['name']}")
                continue
                
            tourney_id = t_record[0]['id'] if isinstance(t_record, list) else t_record.get('id')

            # 2. Sync Teams
            opponents = match.get('opponents', [])
            if len(opponents) != 2:
                continue # Skip if not a 1v1 team match yet (TBD)
                
            team_ids = []
            for opp in opponents:
                team = opp['opponent']
                team_data = {
                    "name": team['name'],
                    "slug": team['slug'],
                    "logo_url": team.get('image_url')
                }
                # Upsert team
                supabase_request("POST", "teams?on_conflict=slug", payload=team_data)
                
                # Get local team ID
                t_json = supabase_request("GET", f"teams?slug=eq.{team['slug']}&select=id")
                if t_json and len(t_json) > 0:
                    local_team_id = t_json[0]['id']
                    team_ids.append(local_team_id)
                    
                    # Add team to tournament participants
                    supabase_request("POST", "tournament_teams?on_conflict=tournament_id,team_id", payload={"tournament_id": tourney_id, "team_id": local_team_id})
                else:
                    print(f"⚠️ Could not find or create team: {team['name']}")

            if len(team_ids) != 2:
                continue

            # Extract scores safely
            results_list = match.get('results') or []
            score1 = results_list[0].get('score', 0) if len(results_list) > 0 else 0
            score2 = results_list[1].get('score', 0) if len(results_list) > 1 else 0

            # 3. Create/Update Match
            match_data = {
                "tournament_id": tourney_id,
                "team1_id": team_ids[0],
                "team2_id": team_ids[1],
                "score1": score1,
                "score2": score2,
                "round": "Upcoming",
                "match_date": match['begin_at']
            }
            
            # Upsert match (we don't have a great unique key here easily, so we just append for now or match on team1/team2/tourney)
            # In a real scenario you'd add a remote_id column to matches to do true upsert
            supabase_request("POST", "matches", payload=match_data)
            print(f"Added Match: {match['name']}")

    print("\nPandaScore Sync Complete!")

if __name__ == "__main__":
    fetch_and_sync_matches()

File: src/lib/ingestion/engines/PandaScoreEngine.js
/**
 * KhelPediA Data Ingestion Engine — PandaScore API
 * Ingests tournaments, teams, and matches across CS2, Valorant, Dota 2, and LoL.
 * Implements complete pagination, exponential backoff retries, and idempotent source mapping.
 */

import { BaseFetcher } from '../BaseFetcher.js';
import { BasePaginator } from '../BasePaginator.js';
import { DataValidator } from '../DataValidator.js';
import { IngestionLogger } from '../IngestionLogger.js';
import { IdempotentUpsert } from '../IdempotentUpsert.js';
import { supabase } from '../../supabase.js';

const PANDASCORE_BASE_URL = 'https://api.pandascore.co';

export class PandaScoreEngine {
  constructor(apiKey = process.env.PANDASCORE_API_KEY) {
    this.apiKey = apiKey ? apiKey.trim().replace(/^["']|["']$/g, '') : null;
    this.logger = new IngestionLogger('pandascore', 'SyncMatches');
    this.fetcher = new BaseFetcher({
      userAgent: 'KhelPediA-PandaScoreEngine/2.0',
      logger: this.logger
    });
  }

  getHeaders() {
    return {
      'Authorization': `Bearer ${this.apiKey}`,
      'Accept': 'application/json'
    };
  }

  async run() {
    console.log("🚀 Starting PandaScore Ingestion Engine...");
    if (!this.apiKey) {
      console.warn("⚠️ PANDASCORE_API_KEY is not set. Skipping PandaScore ingestion.");
      this.logger.logRequest(false, false);
      this.logger.printReport();
      return;
    }

    // 1. Map database games
    const { data: games } = await supabase.from('games').select('id, slug');
    const gameMap = {};
    if (games) {
      games.forEach(g => {
        const slug = g.slug.replace('cs2', 'csgo');
        gameMap[slug] = g.id;
        gameMap[g.slug] = g.id;
      });
    }

    const supportedGames = ['valorant', 'csgo', 'dota-2', 'league-of-legends'];

    for (const gameSlug of supportedGames) {
      const dbGameId = gameMap[gameSlug];
      if (!dbGameId) continue;

      console.log(`\n--- Fetching PandaScore ${gameSlug.toUpperCase()} Matches ---`);

      // 2. Fetch matches using complete pagination
      const matches = await BasePaginator.paginatePandaScore({
        fetchFn: async (page, pageSize) => {
          const url = `${PANDASCORE_BASE_URL}/${gameSlug}/matches?page[number]=${page}&page[size]=${pageSize}&sort=-begin_at`;
          return await this.fetcher.fetch(url, { headers: this.getHeaders() });
        },
        pageSize: 100,
        maxPages: 10, // Fetch up to 1,000 matches per game
        logger: this.logger
      });

      console.log(`Processing ${matches.length} ${gameSlug} matches...`);

      for (const m of matches) {
        try {
          const sourceMatchId = m.id;
          const tourneyObj = m.tournament || {};
          const leagueObj = m.league || {};

          // Validate opponent count
          const opponents = m.opponents || [];
          if (opponents.length !== 2) {
            this.logger.logSkipped();
            continue; // Skip FFA or TBD non-binary team matches
          }

          // A. Upsert Tournament
          const tourneyName = `${leagueObj.name || ''} ${tourneyObj.name || ''}`.trim() || m.name || 'Unknown Tournament';
          const tourneySourceId = tourneyObj.id || m.id;

          const tourneyId = await IdempotentUpsert.upsertTournament({
            source: 'pandascore',
            source_tournament_id: tourneySourceId,
            name: tourneyName,
            game_id: dbGameId,
            region: tourneyObj.region || leagueObj.name || 'Global',
            status: m.status === 'finished' ? 'completed' : (m.status === 'running' ? 'live' : 'upcoming'),
            prize_pool: 0,
            tier: (tourneyObj.tier || 'b').toUpperCase(),
            start_date: tourneyObj.begin_at || m.begin_at,
            end_date: tourneyObj.end_at || m.end_at,
            logger: this.logger
          });

          if (!tourneyId) continue;

          // B. Upsert Teams
          const team1Obj = opponents[0]?.opponent || {};
          const team2Obj = opponents[1]?.opponent || {};

          const team1Id = await IdempotentUpsert.upsertTeam({
            source: 'pandascore',
            source_team_id: team1Obj.id,
            name: team1Obj.name,
            logo_url: team1Obj.image_url,
            logger: this.logger
          });

          const team2Id = await IdempotentUpsert.upsertTeam({
            source: 'pandascore',
            source_team_id: team2Obj.id,
            name: team2Obj.name,
            logo_url: team2Obj.image_url,
            logger: this.logger
          });

          if (!team1Id || !team2Id) continue;

          // Resolve scores & winner
          const results = m.results || [];
          const score1 = results[0]?.score || 0;
          const score2 = results[1]?.score || 0;
          let winnerId = null;
          if (m.winner_id === team1Obj.id) winnerId = team1Id;
          if (m.winner_id === team2Obj.id) winnerId = team2Id;

          // C. Validate Match object
          const matchPayload = {
            source_match_id: sourceMatchId,
            team1_name: team1Obj.name,
            team2_name: team2Obj.name,
            tournament_name: tourneyName,
            score1,
            score2,
            played_at: m.begin_at
          };

          const validation = DataValidator.validateMatch(matchPayload);
          if (!validation.valid) {
            this.logger.logValidationFailure();
            continue;
          }

          // D. Idempotent Match Upsert via entity_source_mapping
          await IdempotentUpsert.upsertMatch({
            source: 'pandascore',
            source_match_id: sourceMatchId,
            tournament_id: tourneyId,
            team1_id: team1Id,
            team2_id: team2Id,
            score1,
            score2,
            winner_id: winnerId,
            round: m.name || 'Group Stage',
            played_at: m.begin_at,
            raw_payload: m,
            logger: this.logger
          });

        } catch (err) {
          console.error(`Error processing PandaScore match ${m.id}:`, err.message);
        }
      }
    }

    await this.logger.printReport();
  }
}
